Generate and export API

kevinrizza opened the pull request:
Added courier_api.py to stub out api methods
Renamed package from command to operatorcourier for readability
on import
Updated cli.py to call api methods
Used argparse to document new cli tool methods that call api
Updated readme documentation with new uses
Added test case file for api

MartinBasti commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,63 @@
+        if not hasattr(self, args.command):
+            print('Unrecognized command')
+            parser.print_help()
+            exit(1)
```
Please do `parser.error('error text')`

kevinrizza replied:
Addressed in latest commit.

MartinBasti replied:
with parser.error you don't need exit(1) or print_help, it will exit with SystemExit=2

MartinBasti replied:
usage text will be printed automatically with parser.error

MartinBasti commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,63 @@
+        parser.add_argument('command', help='Subcommand to run')
+
+        args = parser.parse_args(sys.argv[1:2])
+        if not hasattr(self, args.command):
```
This can be done with subparsers https://docs.python.org/2/library/argparse.html#sub-commands

kevinrizza replied:
I did read about those, but I thought the pattern presented here was very readable and didn't seem to come with any specific disadvantages. What is the value of using subparsers rather than the current pattern?

MartinBasti replied:
I prefer to use stdlib solutions

kevinrizza replied:
Isn't this part of the stdlib? https://docs.python.org/3/library/functions.html#getattr

If there's an obvious standard I'm not following, I can modify the source to follow it. But from my reading it wasn't clear that there was a pythonic standard to add commands to a cli tool. Is there some documentation you can point me to that shows using subcommand as a standard? 

I did take a look at some implementations and it seems like using subcommand here would prevent me from using the pattern here (creating separate methods for each subcommand) which to me is very readable.

MartinBasti replied:
Subparsers are standard way how to add subcommands, but let leave this as it is now

MartinBasti commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,63 @@
+
+        args, leftovers = parser.parse_known_args(sys.argv[2:])
+        if args.files is None and args.source_dir is None:
+            print("Neither source_dir nor files is set. One (and exactly one) of those is required to create a bundle from.")
```
`parser.error('...')`

kevinrizza replied:
Addressed in latest commit.

MartinBasti commented on operatorcourier/courier_api.py:
```diff
@@ -0,0 +1,36 @@
+from operatorcourier.build import BuildCmd
+from operatorcourier.validate import ValidateCmd
+
+def build_and_verify(source_dir="", files=""):
```
Is better to use `None` as default vaules instead of empty string

kevinrizza replied:
Addressed in latest commit.

MartinBasti commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,63 @@
+def main():
+    """Generate the CLI bits
+    """
+    _CliParser()
```
This looks weird to me. Object init should not do any exec operations, it should just create object.

```python
parser = _CliParser()  # just prepares parser
args = parser.parse()  # process args, can do some post parse checks

if args.command == 'push':
    do_push(args.x, args.y)
elif args.command == 'build':
   do_build(args.x, args.z)
else:
  raise RuntimeError('Unexpected commnad')
```

kevinrizza replied:
Addressed in latest commit.

MartinBasti commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,63 @@
+from operatorcourier import courier_api
```
Python module import order should be

```
stdlib imports

3rd party libs import

local app imports
```

```python
import argparse
import sys

from operatorcourier import courier_api
```

kevinrizza replied:
Addressed in latest commit.

MartinBasti commented on operatorcourier/courier_api.py:
```diff
@@ -0,0 +1,36 @@
+"""
+operatorcourier.courier_api module
```
IMHO `operatorcourier.api` is enough, is is clear form import path that API belongs to operator courier

kevinrizza replied:
Addressed in latest commit.

MartinBasti commented on setup.py:
```diff
@@ -1,14 +1,14 @@
 setup(
   name='operator-courier',
-  packages=['command'],
+  packages=['operatorcourier'],
```
:heart: I wanted to propose name change, but you were faster. It sounds much better.

kevinrizza commented:
@MartinBasti Latest commit addressed comments. The pr I created earlier this morning still needs to be merged and then this pr needs to be rebased again, but it should be good to review for the requested changes.

awgreene commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,56 @@
+    elif files is not None:
+        yamls = files.split(",")
+
+    bundle = BuildCmd().build_bundle(yamls)
```
The CLI suggests that only the source_dir or files should be set. If the API doesn't throw an error, we may want to document that the source_dir has priority.

MartinBasti replied:
why not to combine both inputs ?

kevinrizza replied:
Not sure I follow. They are very different inputs. one is a single string that is a path, the other is a list of serialized yaml. do you mean they should just be one parameter and we have code that determines if it is a path or not?

MartinBasti replied:
No, I want two parameters, just append "files" to yamls via `yaml.extend()` when defined

kevinrizza replied:
@MartinBasti are you planning on using the cli interface or importing the api module? Your team is the only reason we were planning on supporting passing the files themselves rather than a directory parameter, and this method is only written that way to support the cli use case. If you aren't going to use the cli, I will just make this a python list parameter and remove that param from the cli.

awgreene commented on operatorcourier/cli.py:
```diff
@@ -0,0 +1,69 @@
+        parser.add_argument('--files', help='Comma separated list of yaml files to bundle. Either set this or use the path argument for bundle data.')
+
+        args, leftovers = parser.parse_known_args(sys.argv[2:])
+        if args.files is None and args.source_dir is None:
```
If only files or source_dir can be set, shouldn't we check if both are set? If this check isn't added, we may want to add the source_dir has greater priority

awgreene commented on tests/test_files/valid_bundles/results/bundle.2.yaml:
```diff
@@ -0,0 +1,206 @@
+data:
```
This file seems to be identical to `tests/test_files/valid_bundles/results/bundle.1.yaml`, why does it exist?

kevinrizza commented on README.md:
```diff
@@ -2,6 +2,30 @@
+
+We also support passing a comma separated list of strings that make up the bundle by specifying the `files=` parameter, i.e.:
+
+`api.build_verify_and_push($your_namespace, $your_repository, $your_release_version, $your_quay_authtoken, files="MyYaml: Value1, MyYaml: Value2")`
```
The example just above this one uses the `source_dir` parameter instead of the `files` parameter.

MartinBasti replied:
why CSV for files instead of list?

kevinrizza replied:
@MartinBasti I'm not sure I follow you? It is a list of serialized yaml files separated by commas. Do you mean the command line parameter should be serialized as a python list?

Also I assume you mean comma separated values (CSV) and not ClusterServiceVersion (CSV), we should probably be more specific in this case.

MartinBasti replied:
But why do you use string with comma-separated-values? Isn't easier to use just standard list/tuple instead having complications with parsing string?

Just question, what happen when yaml will contain textual value with comma inside?

MartinBasti replied:
also `files` name is tricky, when I see that, I think it expects filepath or fileobject, not content

MartinBasti commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,56 @@
+
+    with TemporaryDirectory() as temp_dir:
+        with open('%s/bundle.yaml' % temp_dir, 'w') as outfile:
+            yaml.dump(bundle, outfile, default_flow_style=False)
```
to avoid race-conditions, you may need `outfile.flush()` after dump, to make sure that data were successfully written to file before they will be access from new filedescriptor.

kevinrizza replied:
@MartinBasti I thought the TemporaryDirectory() method creates a new, unique directory path? In that case, what race condition could we hit if the file and path are unique?

kevinrizza commented:
@awgreene @MartinBasti latest commit addressed all comments. please review again.

MartinBasti commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,66 @@
+
+    if source_dir is not None: 
+        for filename in os.listdir(source_dir):
+            if filename.endswith(".yaml"):
```
is possible to have `.yml` ending? Both `.yaml` and `.yml` are used with files associated to YAML content

kevinrizza replied:
updated to include them.

MartinBasti commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,66 @@
+                with open(source_dir + "/" + filename) as f:
+                    yaml_files.append(f.read())
+    elif yamls is not None:
+        yaml_files = yamls
```
why not `yaml_files.extend(yamls)` to respect both options?

If options are mutually exclusive, you should check and raise an exception when both options were used.

kevinrizza replied:
IMO I am already concerned about the supportability of passing the files as a list (OLM is already implementing a change to use a nested hierarchy with different crds associated with specific versions of csvs), so I would like to keep them separate. I added a check that prevents both parameters from being passed.

MartinBasti commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,66 @@
+
+    if not valid:
+        bundle = None
+        logger.error("Bundle failed validation.")
```
Shouldn't be this an exception?

kevinrizza replied:
This api is informational, so passing bad files shouldn't be an exception. A log message is the correct output. The push method should test this and throw an exception when validation fails though.

MartinBasti commented on operatorcourier/api.py:
```diff
@@ -0,0 +1,66 @@
+
+            PushCmd().push(temp_dir, namespace, repository, revision, token)
+    else:
+        logger.error("Bundle is invalid. Will not attempt to push.")
```
IMO this should be exception

kevinrizza replied:
Agreed, raising an exception here now.

MartinBasti commented on tests/test_api.py:
```diff
@@ -0,0 +1,32 @@
+("tests/test_files/bundles/api/bundle1", "tests/test_files/bundles/api/bundle1/results/bundle.yaml"),
+])
+def test_make_bundle(dir, expected):
+    bundle = api.build_and_verify(source_dir=dir)
```
`dir` is python's built-in, please use different name for variable

kevinrizza replied:
updated.

kevinrizza commented:
@MartinBasti addressed your comments.
